lib/amigaos.c

```diff
@@ -90,6 +90,45 @@ void Curl_amiga_X509_free(X509 *a)
 {
   X509_free(a);
 }
+
+/* AmiSSL replaces many functions with macros. Curl requires pointer
+ * to some of these functions. Thus, we have to encapsulate these macros.
+ */
+
+#include "warnless.h"
+
+#pragma push_macro("SHA256_Init")
+#undef SHA256_Init
+int SHA256_Init(SHA256_CTX *c)
+{
+#pragma pop_macro("SHA256_Init")
+  return SHA256_Init(c);
+};
+
+#pragma push_macro("SHA256_Update")
+#undef SHA256_Update
+int SHA256_Update(SHA256_CTX *c, const void *data, size_t len)
+{
+#pragma pop_macro("SHA256_Update")
+  return SHA256_Update(c, data, curlx_uztoui(len));
+};
+
+#pragma push_macro("SHA256_Final")
+#undef SHA256_Final
+int SHA256_Final(unsigned char *md, SHA256_CTX *c)
+{
+#pragma pop_macro("SHA256_Final")
+  return SHA256_Final(md, c);
+};
+
+#pragma push_macro("X509_INFO_free")
+#undef X509_INFO_free
+void X509_INFO_free(X509_INFO *a)
+{
+#pragma pop_macro("X509_INFO_free")
+  X509_INFO_free(a);
+};
+
```

Comment:
Why all the `#pragma` magic? Won't it suffice to just undef the macros?

Reply:
After an undef the function definition line 127 would work. However, for line 129 I need the original macro defined again as otherwise `X509_INFO_free()` would be unknown. So after `pop_macro()` I‘ve got back the original definition of it.  

There is another way by defining 

```
void X509_INFO_internal_free(X509_INFO *a);
void X509_INFO_free(X509_INFO *a){
  X509_INFO_internal_free(a);
};
```

**before** including anything, e.g. in line 22. Then I could write at the bottom of the file:

```
void X509_INFO_internal_free(X509_INFO *a){
  X509_INFO_free(a);
};
```

By inserting something in the name now the macro wouldn‘t replace the function name but only the `X509_INFO_free()`. However, that looks somewhat dirty. 

Note that my changes are within `#ifdef USE_AMISSL` which can only be true if compiled for Amiga with gcc so that shouldn‘t fail.

Reply:
The code is specific for AmiSSL sure, but that fact doesn't make us not want the code to be easy to read and use as few "special tricks" as possible...

How are the defines done? Are they done with parentheses? If so, then maybe it could be written like this, without pragma or undefines:
~~~c

int (SHA256_Init)(SHA256_CTX *c)
{
  return SHA256_Init(c);
};